### ISSUE TITLE
PriorityScheduler and preemption

### ISSUE THREAD

**Issue by csimmons0** [threadly/threadly#245]
I have a couple of questions about PriorityScheduler. In my use case, the high-priority tasks finish in seconds, while the low-priority tasks take minutes. I would want high-priority tasks to preempt low-priority tasks. I looked at the source code of PriorityScheduler, and it doesn’t look like it works that way. If I’m not mistaken, threads that handle tasks all run at the same priority, and, if the max number of threads has been reached, a high-priority task will be delayed even if all threads are handling low-priority tasks. Is my understanding accurate?

**Comment by lwahlmeier** [comment]
Partially.  If all threads are processing tasks, when the next thread to run pulls a task it will make a choice to pull either a high-priority task or a low one.  A low one will be pulled first if and only if it has been waiting longer then a high task +low priority delay.  The low priority delay can get seen/changed with the get/setMaxWaitForLowPriority() on the scheduler.  It is important to note that its pretty much not possible to "preempt" a task high/low or otherwise once it starts, it has to finish at that point.  Doing so would require something like green threads and a kind of yield to know when it can switch tasks.

There is another task type called "Starvable" that might be more what your looking for.  Basically if there is anything else to run high/low Starvable will not run, and only run once nothing else is ready in the low/high queue.

Another option I have used sometimes is to create a "ExecutorLimiter" or a "SubmitterSchedulerLimiter"  you wrap your PriorityScheduler in a "DefaultPriorityWrapper" with a low or even starvable priority and limit the number of threads that that are allowed to execute the longer lived tasks so those tasks can never starve the pool.  This is kind of what the "CentralThreadlyPool" does when you create schedulers using the "rangedThreadPool(TaskPriority priority, int guaranteedThreads, int maxThreads)" type methods.

Hope that helps, let us know if you have anymore questions.

**Comment by jentfoo** [comment]
Feel free to re-open or open a new issue if you have any further questions.  We are always happy to explain implementation
